### ISSUE TITLE
Android 12 Support

### ISSUE THREAD

**Issue by manualexSP** [dotintent/react-native-ble-plx#908]
- [ x] I am running the latest version
- [ x] I checked the documentation and found no answer
- [ x] I checked to make sure that this issue has not already been filed
- [ x] I'm sure that question is related to the library itself and not Bluetooth Low Energy or Classic in general. If that so, please post your question on [StackOverflow](https://stackoverflow.com/questions/tagged/react-native-ble-plx?sort=active) or on our [Gitter](https://gitter.im/RxBLELibraries/react-native-ble) channel.

## Question

Hey guys I'm wondering that if you have in plan to add support for Android 12 any time soon, cause I started to test how my app is working on Android 12 and I have some problems.
When I'm disconnecting from the device I'm getting this error:
```
{
message: "Device ......... was disconnected"
errorCode: 201
attErrorCode: 0
iosErrorCode: null
androidErrorCode: null
reason: "Disconnected from ........... with status 0 (GATT_SUCCESS)"
name: "BleError"
}
```
which is good but attErrorCode is always 0 no matter what I do, I tried to go out of range, reset the device but the attErrorCode is not changing. I tested on a physical device Pixel 5 with Android 12 Beta 5. I couldn't replicate this problem on any other device with any other version of android.
I'm asking this question because I saw that Google made some changes around the Bluetooth permissions and I'm wondering if this might affect this library somehow.
Currently, I'm running on:
"react-native": "0.63.3"
"react-native-ble-plx": "^2.0.2"
So, can you please tell me if the Android 12 can impact you guys somehow and if you have anything planned to support Android 12?

**Comment by thespacemanatee** [comment]
Works for me and I'm on Android 12. Nothing much to offer here because it just works as is and I had no idea there were issues.

**Comment by bostnis** [comment]
unless enable ACCESS_FINE_LOCATION, ACCESS_FINE_LOCATION, startDeviceScan returns error. But from Android 12, ACCESS_FINE_LOCATION is not essentially needed.

could anybody provide solution for this?

**Comment by bostnis** [comment]
https://github.com/dariuszseweryn/RxAndroidBle/releases/tag/release-1.13.1-rxjava2

rxandroidble 1.13.1 which supports api 31(android 12), and it would be awesome if it is integrated to this repo as soon as possible.

**Comment by bostnis** [comment]
@manualexSP  did you already make workaround solution by forking this repo?

**Comment by manualexSP** [comment]
@bostnis ok so these are the problems that I had:
1) the new permissions
2) not getting any error code on attErrorCode
3) sometimes the monitor was failing when it was trying to read the services and the characteristics.

These are the fixies that worked for me:
1) I used react-native-permissions to fix the permissions problems. It worked because this library is updated to the latest react-native version and is also offering support for Android 12 permissions. Unfortunately, I still need to request the location permissions because the react-native-ble-plx and react-native-nordic-dfu are not having support for the new permissions on Android 12.
Also as tagrgets and compile versions we have:
``` 
        buildToolsVersion = "30.0.2"
        minSdkVersion = 21
        compileSdkVersion = 30
        targetSdkVersion = 30
```
because react-native-nordic-dfu is crashing if I make the increase to 31 XD.
2) this problem I still didn't manage to fix yet.
3) My flow works like this: I scan and connect to the device after this I was getting all the services and all the characteristics from the device and then I initialize the monitor. After the monitor was initialized I was trying to get all the information from the device and after 30sec the monitor was throwing an error saying that the device was disconnected. This problem is happening only when I was connecting to a device for the first time.
The problem here was with the pairing request pop-up from the system. 
- On android below 12 when you pair a device with the app for the first time you get a system pop-up from android that is asking the user if he wants to pair with the device or not. When this pop-up was appearing the app was "stopping" and it was waiting for the user to choose `Pair` or `Cancel`. And if the user was selecting Pair then the flow was continuing and working as expected. And if the user was selecting Cancel then we were getting a `Cancel pairing` error.
- but on Android 12 the pairing pop-up from the system is appearing only for a few seconds and is not waiting for any decision from the user, and even the user pressing `Cancel` we were not getting any errors only after 30 seconds. And if the user was pressing `Pair`  we were getting an error of disconnecting the device after 30 seconds because the monitor couldn't get any data.

So the problem here I think it is because the pairing pop-up is not sending back in time any response.
In all the cases the connection between the device and phone was successful but the monitor was failing all the time when I was adding a new device.

How I fix this ? by trying to write a Characteristic before the initialization of the monitor -_-"
I don't know why this worked but it worked, for now.

The MultiPlaformBleAdaptor is very far away from RxAndroidBle and unfortunately, I didn't manage to make the fork possible.

Please keep in mind that every "fix" that I did is actually only workarounds and the main problems are the permissions and the fact that sometimes we don't get any attErrorCode or anything useful. I used a lot of await functions and I'm catching and treating all the error.

I hope that this information helps.

**Comment by bostnis** [comment]
Thanks for detailed explanation about your approaches @manualexSP

**Comment by shubhamraka08** [comment]
@manualexSP I am using react-native-permissions to get the permission but before the popup comes for permission my app is getting crashed. so how do you solve this issue.

**Comment by manualexSP** [comment]
@shubhamraka08 sorry but i didn't had any crash.

**Comment by shubhamraka08** [comment]
@manualexSP  ok Thanks I figured out the issue we are initializing ble plx before getting the permission that's why an app is getting crashed.

**Comment by ariccio** [comment]
Is this still a problem? I'd like to know before I bite the bullet and switch to using 12 for development!

**Comment by manualexSP** [comment]
@ariccio well, we don't have any update to support Android 12, and if you didn't find any workaround, I would say yes, is still a problem.

**Comment by ariccio** [comment]
How good are the workarounds? Can you use react-native-blx in production or dev android 12 devices? 

If you have a moment, I'd really appreciate if you'll indulge my concerns *puppy eyes*

I'm a novice RN dev, readying an app for production in the next few weeks. It's not a big shop with a lot of resources, just me :)

I develop on my primary device, a Pixel XL that's really on it's last legs. Like, everything you'd expect from a 6 year old phone that's hard to repair; sound doesn't work, battery has about the capacity of a large capacitor, etc... Can't afford right now to buy an android 11 dev device.

I am switching to a Pixel 6 pro as soon as the supply chain issues resolve. I'm worried, will I be able to use Android 12 to develop, should my OG Pixel die?

If I was a java guy, I'd help and port the package over! But I'm a C++ guy. I've avoided entire career paths to avoid the heresy of Java. :smirk:

I haven't written any in almost ten years. I spent like two whole days trying to get the android toolchain to work with the correct version of Java in my environment. I doubt that @dotintent would want me touching their codebase! :joy:

**Comment by ariccio** [comment]
You know what's funny? It's working fine for me!

**Comment by manualexSP** [comment]
hi @ariccio , can you explain to us if you did something different or how exactly it works for you?
Any information can help.

**Comment by ariccio** [comment]
Honestly, I have no idea! Is the problem only in device disconnection? I don't attempt to disconnect, because I need to play nicely with a 3rd party app that uses the same device.

**Comment by manualexSP** [comment]
@ariccio the only problem left for me is:
- i don't receive att error codes on Android 12
- and the fact that the library is requesting a location permission to scan for device, which on android 12 is not necessary anymore

**Comment by Irenedan** [comment]
> 

So do you mean "react-native-ble-plx" is available on Android12?

**Comment by Irenedan** [comment]
> @ariccio the only problem left for me is:
> 
> * i don't receive att error codes on Android 12
> * and the fact that the library is requesting a location permission to scan for device, which on android 12 is not necessary anymore

@manualexSP So do you mean "react-native-ble-plx" is available on Android12 finally?

**Comment by ankushpatel18** [comment]
**Question**
Hey guys I'm wondering that if you have added support for Android12. I started to update my react-native app to support Android 12 and I have some problems.

- Android 12 don't need location permission while using Bluetooth.
- Android 12 has added extra bluetooth permissions that must be granted by user.

Is there any plan to update [MultiPlatformBleAdapter](https://github.com/dotintent/MultiPlatformBleAdapter) to support android 12?

Please suggest my any solution if anyone has overcame from this issue.

**Comment by fabiendem** [comment]
@mciechanowicz @bor-ski, sorry to tag you here, but we would need some help with this, or some clarity on the maintenance of the repo.

Currently, `react-native-ble-plx` uses a version of https://github.com/dotintent/MultiPlatformBleAdapter which is not wrapping the latest `RxAndroidBle` library.
Instead it wraps an old version of `RxAndroidBle` which does runtime checks on the permissions, requiring the Location permissions on Android 12 (with `targetSdkVersion 31+`) while it's not needed, even if the new Bluetooth Permissions are granted.
The latest version of `RxAndroidBle` doesn't have this issue.

It looks like the update should go in two steps: 
- Update the https://github.com/dotintent/MultiPlatformBleAdapter so it uses the latest `RxAndroidBle` library. (can be tricky because you move from RxJava 1 to 2)
- Then update `react-native-ble-plx` so it uses the updated `MultiPlatformBleAdapter`

Could you please clarify the status of the `react-native-ble-plx` project since `.intent` took ownership?
Should we look at alternative libraries?
Forking and fixing it ourselves could be a solution, but if the project is no longer maintained by a community, and if the fixes never land on the main branch, it may not be a viable long term solution.

Thanks

**Comment by ericschaal** [comment]
Hi @fabiendem and others

That's exactly what we did here [react-native-ble-plx](https://github.com/MY01Inc/react-native-ble-plx)
We updated [MultiPlatformBleAdapter](https://github.com/dotintent/MultiPlatformBleAdapter) to use the latest RxAndroidBle

We also pushed it to [NPM](https://www.npmjs.com/package/@my01/react-native-ble-plx)

Not sure if it's bulletproof, we didn't go as far as using it in production yet but preliminary testing shows that it works well on Android 12.

It's really sad that this library is no longer maintained.

**Comment by fabiendem** [comment]
Hi @ericschaal 

Wow, amazing effort. Few months ago I started to look into the migration of `MultiPlatformBleAdapter` to RxJava2 and bailed out...
Thanks for sharing.

➡️ On our end, we're considering a migration to https://github.com/innoveit/react-native-ble-manager
- It has better community support now.
- The library is much simpler (no `BleAdapter` layer) and less opinionated.
It may handle less gracefully issues specific to manufacturers, which I think [RxAndroidBle](https://github.com/Polidea/RxAndroidBle) tries to handle

Simplicity and community support may become important for the migration to the new RN Architecture.

The API is very similar to `react-native-ble-plx` and we've been able to swap `react-native-ble-plx` with `react-native-ble-manager` without much troubles.  
The observed behavior is the same, but we've haven't shipped this in prod yet.
One difference I see is the granularity in the error reporting: `react-native-ble-plx` is more fine grained. You get better error codes and details with `react-native-ble-plx`.

I will give a further look at your work and maybe we will use it as an intermediate step.

Thanks again for sharing!

**Comment by MaxToyberman** [comment]
@ericschaal I have tried to use your fork but i see
 BleError: Unknown error occurred. This is probably a bug! Check reason property.]
 
 Do i miss something ?

**Comment by VirtuozTM** [comment]
Hello everyone,

I'm contacting you about a problem I can't seem to solve (about, in my opinion, android permissions).

The problem is the following: I have a mobile application developed in React Native in TypeScript language. Let's take the case, where the problem occurs. I am on a screen and I click on a button. This button makes me navigate to another screen and when I open this screen, I ask for different permissions (such as location, bluetooth etc). I did all my tests on my phone which has an android version 10. Everything works perfectly! However, as soon as I use a friend's phone which has an android 12 version, I get an error when I navigate to the screen. The navigation is fine but when asking for permissions either :

the application crashes
the requests are done but the application is totally buggy and I have to restart it.
In conclusion, when the application is initialized, from the first installation, the permission requests break the application.

I asked help Dan (from https://www.youtube.com/@DanRNLab) because he did a lot of video on BLE. He recommended that I follow his video to fix the android 12 problems but I had already done that :( and it didn't solve my problem.

So you are my last chance to find the error.

A video showing the process and photos are attached.

Hoping that you can help me.

https://user-images.githubusercontent.com/105630116/214044865-de2beff8-1426-4687-8eb8-b841e29b60f0.mp4

![erreur_bluetooth_connect](https://user-images.githubusercontent.com/105630116/214044907-19eadee4-0ed5-4d5d-ad6a-40ee0f630bab.jpg)
![code_permissions1](https://user-images.githubusercontent.com/105630116/214045034-781907f2-0d99-493e-9d1e-694dbd0746bd.jpg)
![code_permissions2](https://user-images.githubusercontent.com/105630116/214045036-11cc8c13-eb1b-4e95-871d-c4e119830203.jpg)

**Comment by VirtuozTM** [comment]
By the way, I don't know if there is a link with the problem but : 
- I use a ` useEffect(() => {
    initBleScanner(serviceUUID);
  }, [initBleScanner, serviceUUID]);` to the screen that is being navigated.
- The function `initBleScanner` looks like : `const initBleScanner = useCallback(
    (UUID_DEVICE: string) => {
      requestPermissions((permissionsGranted: boolean) => {
        if (permissionsGranted) {
          // Code pour continuer avec le scan
          const subscription = manager.current.onStateChange(state => {
            if (state === 'PoweredOn') {
              scanDevices(UUID_DEVICE);
              subscription.remove();
            }
          }, true);
        } else {
          // Code pour gérer l'absence de permission
        }
      });
    },
    [scanDevices],
  );`
- In this function, there is a `requestPermission` function which calls the above function in the attached screenshot.

**Comment by fabiendem** [comment]
@VirtuozTM 🇫🇷  We're maybe missing a piece of the flow here.

In your screenshot:
<img src="https://user-images.githubusercontent.com/105630116/214044907-19eadee4-0ed5-4d5d-ad6a-40ee0f630bab.jpg" width="400" />

It mentions [ACTION_REQUEST_ENABLE](https://developer.android.com/reference/android/bluetooth/BluetoothAdapter#ACTION_REQUEST_ENABLE).
> Activity Action: Show a system activity that allows the user to turn on Bluetooth.

> For apps targeting [Build.VERSION_CODES#S](https://developer.android.com/reference/android/os/Build.VERSION_CODES#S) or or higher, this requires the [Manifest.permission#BLUETOOTH_CONNECT](https://developer.android.com/reference/android/Manifest.permission#BLUETOOTH_CONNECT) permission which can be gained with [Activity.requestPermissions(String[], int)](https://developer.android.com/reference/android/app/Activity#requestPermissions(java.lang.String[],%20int)).
Requires [Manifest.permission.BLUETOOTH_CONNECT](https://developer.android.com/reference/android/Manifest.permission#BLUETOOTH_CONNECT)

➡️ It looks like you're trying to prompt the user to enable the Bluetooth on the device before the `BLUETOOTH_CONNECT` permission got granted. 
➡️ You can validate this and block the reproduction by going into the App System settings and granting all the Bluetooth Permission before you first launch the app. If the error disappears, it's probably this.
➡️ Which `targetSdkVersion` are you using?

**Comment by VirtuozTM** [comment]
Thanks for you response. I dont understand why my code in not correct... The `useEffect` to the screen that is being navigated call a function `initBleScanner`. The main goal of this function is to scan the devices at the initalisation but before to do that, I have to ask permissions to the user especially `BLUETOOTH_CONNECT, BLUETOOTH_SCAN, ACCESS_FINE_LOCATION`. So all permissions are required at the same time. I'm not sure to figure out what you mean. 

"You can validate this and block the reproduction by going into the App System settings and granting all the Bluetooth Permission before you first launch the app. If the error disappears, it's probably this". I already try this one, when I'm going in the settings of my application on the phone. "Authorizations" are disabled and I can active them manually (as you can see on the screenshot just under). Its in french
![disabled_actived](https://user-images.githubusercontent.com/105630116/214234276-d57d7d6c-91e8-4246-8986-86a5bf8c53d0.jpeg)

I'm using `targetSdkVersion = 31`
![SdkVersion](https://user-images.githubusercontent.com/105630116/214235387-03368f43-1d94-4844-be27-753fcb939d62.PNG)

**Comment by abursztynowska** [comment]
I'm excited to inform you that we've released a new version of `react-native-ble-plx`. Many of the concerns and bugs have been addressed in this update.

If you encounter any issues or have further suggestions, please don't hesitate to open a new issue.
